lib/ret/slack_client.ex

```diff
@@ -0,0 +1,117 @@
+defmodule Ret.SlackClient do
+  @oauth_scope "identity.basic,identity.email"
+  @slack_api_base "https://slack.com/"
+
+  def get_oauth_url(hub_sid) do
+    authorize_params = %{
+    response_type: "code",
+    client_id: module_config(:client_id),
+    user_scope: @oauth_scope,
+    state: Ret.OAuthToken.token_for_hub(hub_sid),
+    redirect_uri: get_redirect_uri()
+    }
+
+    "#{@slack_api_base}/oauth/v2/authorize?" <> URI.encode_query(authorize_params)
+  end
+
+  #
+  def fetch_access_token(oauth_code) do
+    body = {
+    :form,
+    [
+      client_id: module_config(:client_id),
+      client_secret: module_config(:client_secret),
+      # grant_type: "authorization_code",
+      code: oauth_code,
+      redirect_uri: get_redirect_uri(),
+      # scope: @oauth_scope
+    ]
+    }
+    %{"authed_user" => authed_user} = "#{@slack_api_base}/api/oauth.v2.access"
+    |> Ret.HttpUtils.retry_post_until_success(body, [{"content-type", "application/x-www-form-urlencoded"}])
+    |> Map.get(:body)
+    |> Poison.decode!()
+
+    Map.get(authed_user, "access_token")
+  end
+
+  def fetch_user_info(access_token) do
+    %{"user" => user} = "#{@slack_api_base}/api/users.identity?" <> URI.encode_query(%{token: access_token})
+    |> Ret.HttpUtils.retry_get_until_success()
+    |> Map.get(:body)
+    |> Poison.decode!()
+
+    user |> Map.put("verified", true)
+  end
+
+
+@permissions %{
+  manage_channels: false,
+  kick_members: false,
+  view_channel: false
+}
+
+  def has_permission?(nil, _, _), do: true
+
+  def has_permission?(%Ret.OAuthProvider{} = oauth_provider, %Ret.HubBinding{} = hub_binding, permission) do
+    oauth_provider.provider_account_id |> has_permission?(hub_binding, permission)
+  end
+
+  def has_permission?(provider_account_id, %Ret.HubBinding{} = hub_binding, permission) do
+    permissions = get_permissions(hub_binding.channel_id, provider_account_id)
+
+    permissions[permission]
+  end
+
+  defp get_permissions(channel_id, provider_account_id) do
+    # Specific channel permissions
+    is_member = is_member_in_channel(channel_id, provider_account_id)
+
+    %{@permissions |
```

Comment:
nit: probably don't need to use this update syntax, since you're updating all the keys and not just some of them. It also means that you don't need `@permissions` at all.